src/Orleans.Core/Messaging/MessageSerializer.cs

```diff
@@ -40,27 +34,17 @@ internal sealed class MessageSerializer : IMessageSerializer
         private object? _bufferWriter;
 
         public MessageSerializer(
-            Serializer<object> bodySerializer,
             SerializerSessionPool sessionPool,
             SharedMemoryPool memoryPool,
-            IServiceProvider services,
-            Serializer<GrainAddress> activationAddressSerializer,
-            ICodecProvider codecProvider,
             MessagingOptions options)
         {
-            _readerSiloAddressCodec = new CachingSiloAddressCodec();
-            _writerSiloAddressCodec = new CachingSiloAddressCodec();
-            _readerIdSpanCodec = new CachingIdSpanCodec();
-            _writerIdSpanCodec = new CachingIdSpanCodec();
-            _serializer = ActivatorUtilities.CreateInstance<Serializer>(services);
-            _activationAddressCodec = activationAddressSerializer;
             _serializationSession = sessionPool.GetSession();
             _deserializationSession = sessionPool.GetSession();
             _memoryPool = memoryPool.Pool;
-            _bodySerializer = bodySerializer;
             _maxHeaderLength = options.MaxMessageHeaderSize;
             _maxBodyLength = options.MaxMessageBodySize;
-            _requestContextCodec = OrleansGeneratedCodeHelper.GetService<DictionaryCodec<string, object>>(this, codecProvider);
+            _requestContextCodec = OrleansGeneratedCodeHelper.GetService<DictionaryCodec<string, object>>(this, sessionPool.CodecProvider);
+            _activationAddressCodec = OrleansGeneratedCodeHelper.GetService<IFieldCodec<GrainAddress>>(this, sessionPool.CodecProvider);
```

Comment:
I wonder if it's worth using partial classes to emit codecs of certain types so that they can be summoned directly and accessed without interface dispatch.

Reply:
A specialized `Dictionary<String, Object>` codec may be worthwhile, too.

Reply:
Probably not worth the effort here, guarded devirtualization will hopefully optimize these callsites.